src/esp/gfx/CubeMap.cpp

```diff
@@ -201,64 +265,84 @@ void CubeMap::recreateTexture() {
+        rgbaAttachment, texture(TextureType::Color), cubeMapCoord,
+        int(mipLevel));
+  }
 
-    if (flags_ & Flag::DepthTexture) {
-      frameBuffer_[index].attachCubeMapTexture(
-          Mn::GL::Framebuffer::BufferAttachment::Depth,
-          texture(TextureType::Depth), cubeMapCoord, 0);
-    } else {
-      frameBuffer_[index].attachRenderbuffer(
-          Mn::GL::Framebuffer::BufferAttachment::Depth,
-          optionalDepthBuffer_[index]);
-    }
+  if (flags_ & Flag::VarianceShadowMapTexture) {
+    frameBuffer_[cubeSideIndex].attachCubeMapTexture(
+        vsmAttachment, texture(TextureType::VarianceShadowMap), cubeMapCoord,
+        int(mipLevel));
+  }
 
-    if (flags_ & Flag::ObjectIdTexture) {
-      frameBuffer_[index].attachCubeMapTexture(
-          objectIdAttachment, texture(TextureType::ObjectId), cubeMapCoord, 0);
-    }
+  // does NOT make any sense to talk about mip level for depth or object id
+  // texture. so the mipLevel is always 0 for both.
+  if (flags_ & Flag::DepthTexture) {
+    frameBuffer_[cubeSideIndex].attachCubeMapTexture(
+        Mn::GL::Framebuffer::BufferAttachment::Depth,
+        texture(TextureType::Depth), cubeMapCoord, 0);
+  } else {
+    frameBuffer_[cubeSideIndex].attachRenderbuffer(
+        Mn::GL::Framebuffer::BufferAttachment::Depth,
+        optionalDepthBuffer_[cubeSideIndex]);
+  }
+
+  if (flags_ & Flag::ObjectIdTexture) {
+    frameBuffer_[cubeSideIndex].attachCubeMapTexture(
+        objectIdAttachment, texture(TextureType::ObjectId), cubeMapCoord, 0);
   }
 }
 
 void CubeMap::prepareToDraw(unsigned int cubeSideIndex,
-                            RenderCamera::Flags renderCameraFlags) {
+                            RenderCamera::Flags renderCameraFlags,
+                            unsigned int mipLevel) {
+  mipLevelSanityCheck("CubeMap::prepareToDraw():", flags_, mipLevel,
+                      mipmapLevels_);
+
+  if ((flags_ & Flag::ManuallyBuildMipmap) && (flags_ & Flag::ColorTexture)) {
+    int size = imageSize_ / pow(2, mipLevel);
+    frameBuffer_[cubeSideIndex].setViewport({{}, {size, size}});
+    recreateFramebuffer(cubeSideIndex, size);
+    attachFramebufferRenderbuffer(cubeSideIndex, mipLevel);
+  }
+
+  bindFramebuffer(cubeSideIndex);
+
   // Note: we ONLY need to map shader output to color attachment when necessary,
   // which means in depth texture mode, we do NOT need to do this
   if (flags_ & CubeMap::Flag::ColorTexture ||
+      flags_ & CubeMap::Flag::VarianceShadowMapTexture ||
       flags_ & CubeMap::Flag::ObjectIdTexture) {
     mapForDraw(cubeSideIndex);
   }
 
   if (flags_ & CubeMap::Flag::ColorTexture &&
       renderCameraFlags & RenderCamera::Flag::ClearColor) {
     frameBuffer_[cubeSideIndex].clearColor(
-        0,                      // color attachment
-        Mn::Color4{0, 0, 0, 1}  // clear color
+        0,                              // color attachment
+        Mn::Color4{0.0, 0.0, 0.0, 1.0}  // clear color
```

Comment:
If you go the extra way, why not extra extra :D

```suggestion
        Mn::Color4{0.0f, 0.0f, 0.0f, 1.0f}  // clear color
```

(I wonder, did the compiler warn here? My assumption is that it shouldn't have for the integer literals, but it should for the double literals?)